libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/OfflineRoute.kt

```diff
@@ -0,0 +1,251 @@
+package com.mapbox.navigation.route.onboard
+
+import androidx.annotation.FloatRange
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.core.exceptions.ServicesException
+import com.mapbox.navigation.route.common.NavigationRoute
+import com.mapbox.navigation.utils.extensions.ifNonNull
+import com.mapbox.navigator.RouterResult
+import okhttp3.HttpUrl
+
+/**
+ * The [OfflineRoute] class wraps the [NavigationRoute] class with parameters which
+ * could be set in order for an offline navigation session to successfully begin.
+ */
+class OfflineRoute
+private constructor(
+    private val onlineRoute: NavigationRoute,
+    bicycleType: OfflineCriteria.BicycleType?,
+    private val cyclingSpeed: Float?,
+    private val cyclewayBias: Float?,
+    private val hillBias: Float?,
+    private val ferryBias: Float?,
+    private val roughSurfaceBias: Float?,
+    waypointTypes: List<OfflineCriteria.WaypointType?>?
+) {
+    private val bicycleType: String?
+    private val waypointTypes: String?
+
+    init {
+        this.bicycleType = bicycleType?.type
+        this.waypointTypes = checkWaypointTypes(waypointTypes)
+    }
+
+    companion object {
+
+        private const val BICYCLE_TYPE_QUERY_PARAMETER = "bicycle_type"
+        private const val CYCLING_SPEED_QUERY_PARAMETER = "cycling_speed"
+        private const val CYCLEWAY_BIAS_QUERY_PARAMETER = "cycleway_bias"
+        private const val HILL_BIAS_QUERY_PARAMETER = "hill_bias"
+        private const val FERRY_BIAS_QUERY_PARAMETER = "ferry_bias"
+        private const val ROUGH_SURFACE_BIAS_QUERY_PARAMETER = "rough_surface_bias"
+        private const val WAYPOINT_TYPES_QUERY_PARAMETER = "waypoint_types"
+
+        /**
+         * Build a new [OfflineRoute] object with the proper offline navigation parameters already setup.
+         *
+         * @return a [Builder] object for creating this object
+         */
+        @JvmStatic
+        fun builder(onlineRouteBuilder: NavigationRoute.Builder): Builder {
+            return Builder(onlineRouteBuilder)
+        }
+    }
+
+    /**
+     * Builds a URL string for offline.
+     *
+     * @return the offline url string
+     */
+    fun buildUrl(): String {
```

Comment:
As discussed we should extract this out into an `interface` and use it across `libdirections-onboard` and `libdirections-offboard`.